src/librustc/hir/map/mod.rs

```diff
@@ -612,7 +618,12 @@ impl<'hir> Map<'hir> {
         if self.dep_graph.is_fully_enabled() {
             let hir_id_owner = hir_id.owner;
             let def_path_hash = self.definitions.def_path_hash(hir_id_owner);
-            self.dep_graph.read(def_path_hash.to_dep_node(DepKind::HirBody));
+            let kind = if hir_id.local_id == ItemLocalId::from_u32_const(0) {
+                DepKind::Hir
+            } else {
+                DepKind::HirBody
+            };
+            self.dep_graph.read(def_path_hash.to_dep_node(kind));
```

Comment:
I don't understand what this logic is supposed to do exactly (before or after the change)...

Reply:
It is supposed to add a `read` to the node which reveals the parent of the `HirId`.

Reply:
OK, I understand the original thinking now: The fingerprint of `HirBody` contains the information of the entire item (including the "signature" parts) so adding a dependency to `HirBody` is the conservative choice. 

But as you discovered the `parent` is not hashed into any of the two dep-nodes, so we are clearly missing dependencies here.

I assume that the `if hir_id.local_id == ItemLocalId::from_u32_const(0)` check is based on the assumption that the parent of any node except the item root is part of the same HIR item and thus the `HirBody` fingerprint contains the appropriate information? If that's the case it is far from obvious and a clarifying comment will be much appreciated by anybody trying to understand this code.